### ISSUE TITLE
autogtp Stop - Resume feature.

### ISSUE THREAD

**Issue by marcocalignano** [gcp/leela-zero#726]
So I finally got an acceptable way to stop and resume autogtp without loosing the games that are running.
It works for Linux (I tested) and I have implemented what should also work for windows. 
So now I need a volunteer that would take my code compile it and test in window. Would be better if this volunteer is willing to work with could read a bit of code and don't mind compiling and running autogtp 100 times if necessary. If anyone want to help me (us actually) just go to my repository and download the branch called /resume

**Comment by l1t1** [comment]
the idea is great, i also do resume with thw windows hiberate function :)

**Comment by marcocalignano** [comment]
@l1t1  did you try the branch?

**Comment by l1t1** [comment]
not yet, I don't have a build system

**Comment by l1t1** [comment]
could you release an exe version?

**Comment by marcocalignano** [comment]
No I cannot build binaries (otherwise I could test it by myself)

**Comment by bood** [comment]
I think you guys can try setting up Appveyor, the built binaries can be downloaded.

**Comment by marcocalignano** [comment]
@bood can you explain a bit more?

**Comment by bood** [comment]
See https://www.appveyor.com/docs/packaging-artifacts for details.
Basically you'll have to specify which binary files to archive.
But the openblas version we're using on CI now is a bit old according to @gcp though

**Comment by marcocalignano** [comment]
@bood still I cannot do anything with this no Windows system in my home!

**Comment by bood** [comment]
You can test it there by mocking some interrupted  games, something like unit tests.

**Comment by zediir** [comment]
I unfortunately can't test this until next weekend. If this issue is still current then I can help.

**Comment by marcocalignano** [comment]
Actually you can use the leelaz binary from the release. You just need Qt5, run qmake and make to build autogtp. So anyone else with Qt5 that want to test?

**Comment by alreadydone** [comment]
@marcocalignano Nice work! Any explanation about the intended behavior? I think it can exit autogtp and save the ongoing game(s) before, and now the "capture keyboard input" thing allows people to stop at any time they want, I suppose? If several games are stored under one directory, hopefully now one autogtp instance can pick up one game to resume without affecting other stored games?

I am concerned because after upgrading to 6x128, CPU cores on BRII takes 160+ seconds for a move and maybe 18 hours for a game, so wasted resources due to time limit is becoming more severe. Lowering playouts could be a solution.

**Comment by marcocalignano** [comment]
Before to exit and save the **not finished** game you needed to pass the parameter` -t timeout` with a timeout that you have to know in advance. Now if you just press _`q`_ and then _`enter`_ autogtp save all running game on the HD and next time is startet resume them.

**Comment by marcocalignano** [comment]
@gcp I tested this feature in Linux and it is working. Should I make a PR so Linux users can enjoy it or do we wait till one windows user try it out and report back?

**Comment by alreadydone** [comment]
@marcocalignano I just tested it; both `-t timeout-in-minutes` and `q+Enter` are working. It would be nice if the old sgf file could be automatically deleted. Moreover, if I run `autogtp -g 2` when there are three `storefile*` files in the directory, two games will be picked up and resumed, but the remaining game will be deleted. It would be nice if it doesn't get deleted.

**Comment by alreadydone** [comment]
Can anyone point to me the place in the source to modify to make autogtp exit after one game is finished? Carbonate and Karst supercomputing systems allow jobs to be submitted to different cores on the same node and to terminate at different times, so it's feasible to play one game per job.

**Comment by marcocalignano** [comment]
@alreadydone Did you tested with windows? 
When autogtp starts read all the store games and keep them in memory so it is ok that it deletes the others.
I also should delete all the file  (the sgf too) if not that is a bug I have to look at it.

to do one game only try the following at thi line https://github.com/marcocalignano/leela-zero/blob/fbe3c9dd5d3a135bb701e0596a360032f842a45f/autogtp/Management.cpp#L158

instead of the `if` block write `m_gamesThreads[index]->doFinish()`
and then after  `m_syncMutex.unlock();` add `wait();`

This should do the trick.

**Comment by bood** [comment]
so supercompuer still with us? I thought you are gone since game rate
drpped a lot fron 2000 to 700 per hour

2018年1月26日 05:18，"Junyan Xu" <notifications@github.com>写道：

**Comment by alreadydone** [comment]
@marcocalignano Unfortunately I don't have the setup to compile it on Windows. I'll try the modifications to the source that you suggest.

If there are three `storefile*` files, then `autogtp -g 2` pick up two of the games to resume, but delete all three files. You are saying that the third game is read into the memory and not lost? But if I then hit `g+Enter`, only two `storefile*` will be created, so the third game is effectively lost in the end. I don't know what will happen if I let it run for longer; maybe after one of the two games is finished, the third game will be resumed?

@bood I think @roy7 has a better idea than me how many games are from the supercomputers. About two days ago I've been using 400+ nodes a lot of which have GPUs, resulting in observed ~2000 games/hour. I think my priority are then temporarily lowered and now only able to use 214 nodes on Carbonate, generating ~80 games/hour.

**Comment by marcocalignano** [comment]
@alreadydone The third game will restart after one of the first two finishes, but you are right, when I quit I have to save all the one on that queue not only the running ones. Thanks for the input.
I also have a look if I can add another command line option like `--single` the produces one game and exit so you do not have to mess with the code. (But I guess I will work in the weekend)

**Comment by alreadydone** [comment]
@marcocalignano I would prefer that the games that can't be resumed immediately be written back to disk immediately, so that it can be resumed by another `autogtp` instance started later. Otherwise, the stored games will accumulate and never get cleared up, and a lot of outdated games will continue to be uploaded long after a network update.

By the way, how can I disable input capturing in the source code? I ask because when I use PBS (Portable Batch System) scripts to submit jobs to the supercomputers, I find that there are ~16000 "captured input" message generated every second in the command line output! Do you have an idea what may be causing this?

**Comment by marcocalignano** [comment]
@alreadydone I do not know what you mean with "captured input".

**Comment by alreadydone** [comment]
@marcocalignano When you press `q` and then `Enter`, the message "captured input: q" will appear. The problem is that when I run autogtp with PBS, it looks like that someone hit `Enter` 16000 times every second. I think input capturing is introduced in https://github.com/gcp/leela-zero/commit/6d892035fa3ad2b37fb72c428550f93f48a2788b and I wonder if there is anyway to disable it.

**Comment by marcocalignano** [comment]
@alreadydone if you disable it I do not get the `q` and `Enter` anymore. I guess I can disable it if the option `--timeout` or later `--single` is given. Could that help you?

**Comment by alreadydone** [comment]
@marcocalignano Yes. `q+Enter` is certainly useful for many people, but I do not need it, so it would be nice to provide an option to switch it off.

I modified the source as you suggested; I did see that autogtp stopped playing games after one game, displaying the message "Program Ends: exiting" in the end, but it seems that autogtp didn't terminate itself. Should I not add the `wait();`?

**Comment by zediir** [comment]
Quitting by `q+Enter` or `-t/--timeout` and resuming works fine but `-s/--single` does nothing

```
D:\My Folders\Documents\leela-zero-resume\>autogtp.exe -u1 -s
...
Starting thread 1 on GPU 0
Worker: before execute
5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9    -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 2638149397166723429  --gpu=1  -g -q -w
./leelaz.exe   -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 2638149397166723429  --gpu=1  -g -q -w 5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9
Engine has started.
Infinite thinking time set.
Loading 8bc83f671ae349c58e68190885b04aec.sgf
36 (W Q3) 37 (B R4) 38 ... 444 (W resign) Game has ended.
Score: B+Resign
Winner: black
ProductionJob: Stopping engine.
Worker: after execute
Uploading game: 3cf9c947ef6b4212ad5f79275379fac2.sgf for network 5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9
Game data 56edc99645d95189f9269de2c46207230aaa64ba63d435a8d60e77a4600f7413 stored in database
1 game(s) (1 self played and 0 matches) played in 11 minutes = 682 seconds/game, 1667 ms/move, last game took 682 seconds.
Worker: before execute
5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9    -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 2638149397166723429  --gpu=1  -g -q -w
./leelaz.exe   -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 2638149397166723429  --gpu=1  -g -q -w 5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9
Engine has started.
Infinite thinking time set.
Loading 8bc83f671ae349c58e68190885b04aec.sgf
36 (W Q3) 37 (B R4) ... 45 (B R11) captured input:
^C
D:\My Folders\Documents\leela-zero-resume>
```

**Comment by zediir** [comment]
Oh. Noticed that it plays the same game again with the same random seed

```
5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9   -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 1280421161062143962  --gpu=1  -g -q -w
./leelaz.exe  -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 1280421161062143962  --gpu=1  -g -q -w 5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9
Engine has started.
Infinite thinking time set.
1 (B Q4) ... 388 (W resign) Game has ended.
Score: B+Resign
Winner: black
ProductionJob: Stopping engine.
Worker: after execute
Uploading game: c8282de6563940adb5613510a378b2e4.sgf for network 5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9
Game data 0382a5163833a565494a6f54ec19ab3e45e1294ab04ee7fa052a440257bf3bbe stored in database
1 game(s) (1 self played and 0 matches) played in 10 minutes = 650 seconds/game, 1675 ms/move, last game took 649 seconds.
Worker: before execute
5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9   -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 1280421161062143962  --gpu=1  -g -q -w
./leelaz.exe  -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 1280421161062143962  --gpu=1  -g -q -w 5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9
```

**Comment by marcocalignano** [comment]
@alreadydone Now -t and -s disable the input console so you should be able to use them. Give them a try please.

**Comment by marcocalignano** [comment]
Is there also anyone that can test this branch on Mac?

**Comment by zediir** [comment]
-t1 worked properly.

-s, now manages to stop the engine but after stopping the engine it stored the game even though it already ended.

```
ProductionJob: Stopping engine.
Worker: after execute
Uploading game: af8045e2a0cc4fad891ccda7903afa23.sgf for network 5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9
Game data 8f6f7c9f2f6f7a409524b458efadad9ed7284677a0ab987ce89ff939c702ddfa stored in database
1 game(s) (1 self played and 0 matches) played in 7 minutes = 424 seconds/game, 1669 ms/move, last game took 423 seconds.
Worker: before execute
Management: Signal received saving game(s)
5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9   -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 6749102091119115137  --gpu=1  -g -q -w
Worker: Command receive store my game
Job is saving game
./leelaz.exe  -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 6749102091119115137  --gpu=1  -g -q -w 5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9
Management: waiting for workers
Engine has started.
Infinite thinking time set.
1 (B D4) ProductionJob: Program ends: Storing game.
ProductionJob: Stopping engine.
Worker: after execute
Writing Order
Worker: Program ends: exiting.
Management: Worker 1 ended
```

When I start the engine again with -s I get
```
Starting thread 1 on GPU 0
Worker: before execute
5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9    -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 6749102091119115137  --gpu=1  -g -q -w
./leelaz.exe   -p 1600  -r 1  -m 30  -t 1  -d   -n   --noponder  -s 6749102091119115137  --gpu=1  -g -q -w 5773f44c0f0e690325b3c09c0c07874e835ad885f4fc6f22ecc21cd2a1bf5af9
Engine has started.
Infinite thinking time set.
Loading 9ced3c69904947c0a7e6128336d440d3.sgf
2 (W D16) 3 (B Q4) 4 (W Q16) 5 (B Q6) 6 (W C3) 7 (B D3) 8 (W C4) 9 (B C5) 10 (W B5) 11 (B C6) 12 (W B6) 13 (B C7)
```
Which is the same job it just finished. (Both had the same -s 6749102091119115137 )

Stopping and starting with `q+enter` still works fine without the -s or -t parameters.

**Comment by marcocalignano** [comment]
On my computer works but I know what the problem could be. Thread
scheduling could make the difference here try doing `gameTread->doFinish()`
before the `sendQuit()` line. I am already out so I will look at it
tomorrow.

**Comment by zediir** [comment]
+1 That worked. I'll do another pull request with the fix.

**Comment by alreadydone** [comment]
@marcocalignano Disabling the input console with -t or -s works. Still I hope you consider my suggestion https://github.com/gcp/leela-zero/issues/726#issuecomment-360695057 above; if not then it's not of practical value of Big Red II, since it will end up resume only one game in several hundred unfinished games. I don't know whether `-s` is working properly now, but I was able to do one game and exit by replacing `wait();` by `exit(0);` in your suggestion; the system will take care of cleanup if necessary.

**Comment by zediir** [comment]
Tested and something like this seems to work

```
void Management::checkStoredGames() {
    QDir dir;
    QStringList filters;
    filters << "storefile*.bin";
    dir.setNameFilters(filters);
    dir.setFilter(QDir::Files | QDir::NoSymLinks);

    QFileInfoList list = dir.entryInfoList();
    for (int i = 0; i < list.size() && i < m_games * m_gpus; ++i) {
        QFileInfo fileInfo = list.at(i);
        Order o;
        o.load(fileInfo.fileName());
        m_storedOrders.push_back(o);
        QFile::remove(fileInfo.fileName());
    }
}
```

**Comment by marcocalignano** [comment]
@zediir I have another Idea for the last problem that could be better But I will try it later.

**Comment by alreadydone** [comment]
I have two other suggestions for improvements:
(1) allow to specify a directory under which the unfinished games will be stored; I am running a GPU version and a CPU version on BRII and it would be preferable that they can access the same pool of unfinished games, because sometimes more CPUs are available and sometimes more GPUs are available, and you would like the games left by CPUs to be quickly finished by GPUs.
(2) check unfinished games not only at startup but also after each game; this will allow the faster GPUs (if available) clear up the stored games quickly.

**Comment by marcocalignano** [comment]
@alreadydone the second point here is done together with your request to not delete the file untill the game is resume. I am currently testing the code.

**Comment by alreadydone** [comment]
@marcocalignano  It seems that the current autogtp scans the directory to obtain a list of storefile* files only at startup, not after every game. One test I did which revealed a problem is the following:
Suppose there are two stored games in the autogtp directory. If I run autogtp, it will resume one of the stored games and delete one storefile*. Now remove the other storefile* (which may be done by starting another autogtp instance), then after the resumed game is finished, autogtp seems to attempt to resume the other stored game; it doesn't scan the directory again to realize that the game is gone. The following error will then appear: 
```
...
Got previously stored file
Engine has started.
Infinite thinking time set.
Loading 935e103e7a484f44bda6487a4bc96e71.sgf
Loading 935e103e7a484f44bda6487a4bc96e71.train
11 (B F3) ... 278 (W pass) Game has ended.
Score: W+63.5
Winner: white
Uploading game: 359c8e5c44b74dcda65c5ab7107b0536.sgf for network 257aeeb863dc51bfc598838361225459257377a4b2c9abd3e1ac6cdba1fcc88f
Game data 8dd74415e57948f841cbf9fba6a6efe3d7cdcf26c6f44d504c2e28fdb519829b stored in database
1 game(s) (1 self played and 0 matches) played in 14 minutes = 862 seconds/game, 3216 ms/move, last game took 861 seconds.
Got previously stored file
Engine has started.
Infinite thinking time set.
Loading 935e103e7a484f44bda6487a4bc96e71.sgf
GTP: ? cannot load file

*ERROR*: Error in GTP response.
Loading 935e103e7a484f44bda6487a4bc96e71.train
GTP:

*ERROR*: Error in GTP response.
*ERROR*: Error in GTP response.
Error read 3 '=
'
```
Strangely, autogtp tries to load the *.sgf and *.train files that have already be resumed. This problem is the more serious one because it causes autogtp to terminate.

Likewise, currently autogtp will only resume stored games that were already there at its startup. Even after it finishes all these stored games, it won't scan the directory again to see whether any new stored games left by other autogtp instances have appeared in the directory.

So if you really implemented my second point above, it is not working properly unfortunately.

**Comment by marcocalignano** [comment]
I think this I could fix, but autogtp is not meant to be run multiple time in the same directory there are a lot of others problem like the sending pending games, storing SGF or Taining file if the user wants to keep them. So I guess you will fine a lot of this problem. You are the only one using it in this way. Some things I cannot even test them on my machine. So please be patient with me.

**Comment by marcocalignano** [comment]
@alreadydone now fixed in #813

**Comment by Mardak** [comment]
You fixed this with #769 ;)
